Sources/SymDiff/source/BoogieVerify.cs

```diff
@@ -909,28 +910,47 @@ public static int RunVerificationTask(VerificationTask vt, Program prog, out boo
             ReplaceInFile(vt.Eq.Name + "_out.bpl", "@", "_");
             if (!wrapper)
             {
-                prog = BoogieUtils.ParseProgram("RS" + vt.Eq.Name + "_out.bpl");
+                var rs_filename = "RS" + vt.Eq.Name + "_out.bpl";
 
-                if (prog == null)
+                // JATIN_NOTE: This code is a hack to fix all boogie AST bugs. It dumps the program to a file,
+                // and calls boogie on the file, as if from the command line. This change counters unsound behavior
+                // introduced by manipulating boogie programs without maintaining their invariants, which is hard to do.
+                // A note in VerifyImplemenation describe the unsound behavior in more detail.
+                var options = new CommandLineOptions(TextWriter.Null, new ConsolePrinter())
                 {
-                    Log.Out(Log.Verifier, "Parse Error!!! in   " + vt.Eq.Name);
-                    return 1;
+                    RunningBoogieFromCommandLine = true
+                };
+                options.Parse(["/soundLoopUnrolling", "/inline:assume"]);
+                var resultCache = new VerificationResultCache();
+                var engine = new ExecutionEngine(options, resultCache);
+                if (engine == null) {
+                    Log.Out (Log.Verifier, "BIG ERRROR ENGINE");
                 }
-                if (BoogieUtils.ResolveAndTypeCheckThrow(prog, Options.MergedProgramOutputFile, BoogieUtils.BoogieOptions))
-                    return 1;
-
-                newEq = vt.Eq;
-                newProg = prog;
+                prog = BoogieUtils.ParseProgram(rs_filename);
+                var stringWriter = new StringWriter();
+                var success = engine.ProcessProgram(stringWriter, prog, rs_filename).Result;
+                string output = stringWriter.ToString();
+                var verified = output.Contains("0 errors");
+                if (success && verified) {
+                    vt.Result = VerificationResult.Verified;
+                } else if (success) {
+                    vt.Result= VerificationResult.Error;
+                } else {
+                    vt.Result = VerificationResult.Unknown;
+                }
+                Log.Out(Log.Verifier, "DONE");
+                // newEq = vt.Eq;
+                // newProg = prog;
 
-                var vcgen = InitializeVC(newProg);
-                //SDiff.Boogie.Process.ResolveAndTypeCheck(newProg, "");
-                newDict = SDiff.Boogie.Process.BuildProgramDictionary(newProg.TopLevelDeclarations.ToList());
+                // var vcgen = InitializeVC(newProg);
+                // //SDiff.Boogie.Process.ResolveAndTypeCheck(newProg, "");
```

Comment:
Here and below the additional comment characters are unnecessary.

Reply:
I feel they are temporarily useful, so i am going to leave them in. My goal is to uncomment these things.